Allow capturing the content of Ember textareas - fixes #35

jon-hall opened the pull request:
This allows the content of textareas to be captured in percy snapshots, importantly though **this can only work in browsers where `Ember.TextArea` actually works** (so far it seems to not work in PhantomJS, and has been shown to work as expected in Chrome).

I'm not sure if a note needs adding to the README about this browser-based limitation for textareas, but this PR does at least allow snapshots taken from a suitable browser to capture the textarea content.

fotinakis commented:
Hey @jon-hall Thanks for this! This looks great, one question: do you have a sense for why the tests might still be passing since we're running them in PhantomJS 2.1? Based on your description I'd expect that to fail.

jon-hall commented:
Hi @fotinakis, the test I added was just a copy of the existing `input` test, modified to apply to `textarea`.  From what I could tell that test was just:
- Setting up the potential "issue" (something percy wasn't capturing without doing some additional tweaks to the cloned HTML)
- Verifying the problem exists in the (non-cloned) HTML
- Then taking a percy snapshot

It didn't seem to make any assertions on what percy captures (e.g. using `ember-cli-mirage` to intercept the request percy makes to inspect the data captured etc.), so I assumed the idea was that visual verification was then being done in percy itself for the resulting snapshot.  As such, I'd expect the `ember test` command to pass since all the assertions are checking the DOM prior to percy doing anything; though I'd expect the actual percy snapshot taken to show a `textarea` with `<!-- -->` in it.

Thinking about it though, I have realised that the test passing means my assertion about **exactly why** it doesn't work in PhantomJS is probably incorrect - during investigating this I was never able to pull the `textarea` value back out *from the clone* (having checked every property/attribute/etc. I could think of) and so assumed the `Ember.TextArea` just wasn't working in PhantomJS.  Since the test is passing though, it indicates that the original DOM is updated correctly (e.g. `Ember.TextArea` does actually work in PhantomJS), but that the cloned DOM just no longer contains the data (so the DOM cloning is likely the bit which doesn't work as expected in PhantomJS).

It may be possible to improve this to work on PhantomJS by finding `textarea`s in the clone and pulling their value(s) through from the corresponding original `textarea` instead - though this would require mapping between the two.  I think I'll try seeing if I can get this to work (as full browser support would be much better), and update here whether it does work and/or how much complexity it ends up adding to the code.

jon-hall commented:
Here's the updated version - https://github.com/jon-hall/ember-percy/commit/e677f15198cdea87522fb9b48e27263ea0f9c5fe - which deals with the loss of data during cloning by capturing textarea values from the original and later restoring them to the clone in a way that shows up in snapshots.

It's quite a bit more complex than the original, but shouldn't impact rendering in any way and does appear to get the correct data being pushed from textareas to percy for both PhantomJS and Chrome now.  @fotinakis , let me know your thoughts and I can amend this PR to use that solution if you'd prefer.

jon-hall commented:
Hi @fotinakis, have you had a chance to check that alternative (PhantomJS-compatible) approach?

The other approach I thought of which could have been used to map from the original to the clone was by walking the (theoretically identical) DOM trees in parallel, but I felt the use of transient classes to map was potentially more reliable and/or performant (given the apparent issues cloning may have in browsers such as PhantomJS).

I'm happy to amend in light of any suggestions though, as getting this closed and being able to capture textarea content would be very handy.

fotinakis commented:
Super late reply here, sorry.

I spent some time a while back going through the two options, and I definitely thought the simpler one was better (with the downside that it only works in TextArea compatible browsers). Also I think you were correct that this is because of the DOM clone, not because of browser support. The [jQuery docs](https://api.jquery.com/clone/) for `.clone()` actually say:

> Note: For performance reasons, the dynamic state of certain form elements (e.g., user data typed into textarea and user selections made to a select) is not copied to the cloned elements. 

So, LGTM 👍
